File: solve/lua/quick.lua
require "common" 

function quicksort(a, lo, hi)
    local pivot = a[math.floor((lo + hi)/2)]
    local i, j = lo, hi
    while i < j do
        while a[i] < pivot do i = i + 1 end
        while a[j] > pivot do j = j - 1 end
        if i <= j then
            swap(i, j)
            i = i + 1
            j = j - 1
        end 
    end
    if lo < j then quicksort(a, lo, j) end 
    if hi > i then quicksort(a, i, hi) end 
end

skeleton(function ()
    quicksort(a, 1, n)
end)

File: solve/cpp/quick.cc
#include "common.hh"

void quicksort(int *lo, int *hi) {
    int pivot = *(lo + (hi - lo)/2);
    int *i = lo, *j = hi;
    do {
        while (*i < pivot) ++i;
        while (*j > pivot) --j;
        if (i <= j) {
            swapp(i, j);
            ++i; --j;
        } 
    } while (i < j);
    if (i < hi) quicksort(i, hi);
    if (lo < j) quicksort(lo, j);
}

int main(int argc, char *argv[]) {
    if (init(argc, argv) == 0) {
        quicksort(a, a + n - 1);
        output();
        return 0;
    }
    return 1;
}
